src/backend/distributed/shared_library_init.c

```diff
@@ -357,7 +382,56 @@ _PG_init(void)
 	{
 		DoInitialCleanup();
 	}
-	columnar_init();
+
+	/* ensure columnar module is loaded at the right time */
+	load_file(COLUMNAR_MODULE_NAME, false);
+
+	/*
+	 * Now, acquire symbols from columnar module. First, acquire
+	 * the address of the set options hook, and set it so that we
+	 * can propagate options changes.
+	 */
+	ColumnarTableSetOptions_hook_type **ColumnarTableSetOptions_hook_ptr =
+		(ColumnarTableSetOptions_hook_type **) find_rendezvous_variable(
+			"ColumnarTableSetOptions_hook");
+
+	/* rendezvous variable registered during columnar initialization */
+	Assert(ColumnarTableSetOptions_hook_ptr != NULL);
+	Assert(*ColumnarTableSetOptions_hook_ptr != NULL);
+
+	**ColumnarTableSetOptions_hook_ptr = ColumnarTableSetOptionsHook;
+
+	/*
+	 * Acquire symbols for columnar functions that citus calls.
+	 */
+	void *handle = NULL;
+	ExternColumnarSupportsIndexAM = (ColumnarSupportsIndexAM_type) (void *)
+									load_external_function(COLUMNAR_MODULE_NAME,
+														   "ColumnarSupportsIndexAM",
+														   true, &handle);
+	ExternCompressionTypeStr = (CompressionTypeStr_type) (void *)
+							   load_external_function(COLUMNAR_MODULE_NAME,
+													  "CompressionTypeStr",
+													  true, &handle);
+	ExternIsColumnarTableAmTable = (IsColumnarTableAmTable_type) (void *)
+								   load_external_function(COLUMNAR_MODULE_NAME,
+														  "IsColumnarTableAmTable",
+														  true, &handle);
+	ExternReadColumnarOptions = (ReadColumnarOptions_type) (void *)
+								load_external_function(COLUMNAR_MODULE_NAME,
+													   "ReadColumnarOptions",
+													   true, &handle);
```

Comment:
should we instead use `lookup_external_function` in those later calls ?

```c
/*
.....
 *
 * If filehandle is not NULL, then *filehandle will be set to a handle
 * identifying the library file.  The filehandle can be used with
 * lookup_external_function to lookup additional functions in the same file
 * at less cost than repeating load_external_function.
 */
void *
load_external_function(const char *filename, const char *funcname,
					   bool signalNotFound, void **filehandle)
```

Reply:
Also, what about getting read of `INIT_COLUMNAR_PASSTHROUGH_SYMBOL` and then making use of `lookup_external_function` in later calls based on the handle that we will resolve for the first one (`columnar_handler`) ?

Reply:
Thank you. I misread the API and I thought if I called it multiple times with the handle, it would use it. Fixed.

I'm still using `INIT_COLUMNAR_SYMBOL` because we still need to assign the symbol to something.